File: implementations/node/tests/hosted_invocation/validate.mjs
/** Run the installed Node executable without a major-version allow-list; refuse empty/skipped .NET validation. */
import { spawn } from 'node:child_process';
import fs from 'node:fs';
import path from 'node:path';
import crypto from 'node:crypto';
import { fileURLToPath } from 'node:url';

const here = path.dirname(fileURLToPath(import.meta.url));
const root = path.resolve(here, '../../../..');
const args = process.argv.slice(2);
if (args.some(value => value !== '--node-only') || args.length > 1) {
  throw new Error('Usage: node validate.mjs [--node-only]');
}
const nodeOnly = args.includes('--node-only');

async function run(command, argv, options = {}) {
  return await new Promise((resolve, reject) => {
    const child = spawn(command, argv, { stdio: ['inherit', 'pipe', 'pipe'], ...options });
    const chunks = [];
    child.stdout.on('data', value => { chunks.push(value); process.stdout.write(value); });
    child.stderr.on('data', value => process.stderr.write(value));
    child.once('error', reject);
    child.once('close', (code, signal) => code === 0 ? resolve(Buffer.concat(chunks).toString('utf8')) :
      reject(new Error(`${command} exited with ${code ?? signal}`)));
  });
}

console.log(`Hosted TypeScript compatibility: Node.js ${process.versions.node}`);
const output = await run(process.execPath, ['--test', '--test-reporter=tap', path.join(here, 'test_worker.mjs')]);
function count(name) {
  const match = new RegExp(`^# ${name} (\\d+)\\s*$`, 'm').exec(output);
  if (!match) throw new Error(`Node test summary is missing ${name}.`);
  return Number(match[1]);
}
const nodeResults = { total: count('tests'), passed: count('pass'), failed: count('fail'), skipped: count('skipped'), cancelled: count('cancelled') };
if (nodeResults.total < 38 || nodeResults.passed !== nodeResults.total || nodeResults.failed || nodeResults.skipped || nodeResults.cancelled) {
  throw new Error('Node compatibility validation is empty, incomplete or unsuccessful.');
}
if (nodeOnly) {
  console.log('Node-only validation completed. .NET profile, publication/DAG and policy tests were NOT executed.');
} else {
  const dotnet = process.platform === 'win32' ? 'dotnet.exe' : 'dotnet';
  const testProject = path.join(root, 'implementations', 'dotnet', 'Tests', 'Multiplexed.AI.Tests', 'Multiplexed.AI.Tests.csproj');
  if (!fs.existsSync(testProject)) throw new Error('Run this validator inside the complete repository.');
  const runDirectory = path.join(root, 'implementations', 'dotnet', 'TestResults', 'sdk-closure', 'node-' + crypto.randomUUID());
  fs.mkdirSync(runDirectory, { recursive: true });
  const trxPath = path.join(runDirectory, 'sdk-closure-node-compatibility.trx');
  const filter = 'FullyQualifiedName~Multiplexed.AI.Tests.Runtime.Invocation.Workers.TypeScript|' +
    'FullyQualifiedName=Multiplexed.AI.Tests.Runtime.Invocation.Workers.Policies.AiHostedConcurrencyPolicyRealProcessTests.TypeScript_Process_Executes_A_Custom_Concurrency_Policy';
  const environment = { ...process.env, MULTIPLEXED_NODE_EXECUTABLE: process.execPath };
  console.log(`TRX output: ${trxPath}`);
  await run(dotnet, ['test', testProject, '--filter', filter, '--logger',
    'trx;LogFileName=sdk-closure-node-compatibility.trx', '--results-directory', runDirectory], { env: environment });
  const trx = fs.readFileSync(trxPath, 'utf8');
  // Inspect individual outcomes: global notExecuted counters can hide dynamically skipped theories.
  const rows = [...trx.matchAll(/<(?:\w+:)?UnitTestResult\b[^>]*>/gu)].map(match => match[0]);
  if (rows.length < 76 || rows.some(row => !/\boutcome="Passed"/u.test(row))) {
    throw new Error('The .NET TypeScript suite is incomplete, skipped or failing. Inspect its fresh TRX.');
  }
  for (const required of [
    'Published_Synchronous_Function_Returns_Real_Computed_Data',
    'Unstarted_TypeScript_Function_Executes_Original_Code_After_Republication',
    'Corrupt_Host_Compiler_Fails_Before_Process_Readiness',
    'TypeScript_Process_Executes_A_Custom_Concurrency_Policy'
  ]) {
    if (!rows.some(row => row.includes(required))) throw new Error(`Required real-process proof is missing: ${required}`);
  }
  console.log(`Complete Node/.NET validation passed: ${nodeResults.passed} Node tests; ${rows.length} .NET results. TRX: ${trxPath}`);
}

File: implementations/python/tests/hosted_invocation/validate.py
"""Run the Python suite and opt-in .NET/Python integration without silent skips."""
from __future__ import annotations

import argparse
import os
from pathlib import Path
import shutil
import subprocess
import sys
import tempfile
import xml.etree.ElementTree as ET


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--python-only", action="store_true", help="Run Python only; this does not validate the .NET integration.")
    args = parser.parse_args()
    if sys.implementation.name != "cpython" or sys.version_info[:2] not in ((3, 12), (3, 13)):
        print("Use a stable CPython 3.12.x or 3.13.x interpreter.", file=sys.stderr)
        return 2
    root = Path(__file__).resolve().parents[4]
    tests = Path(__file__).resolve().parent
    result = subprocess.run([sys.executable, "-B", "-m", "unittest", "discover", "-s", str(tests),
                             "-p", "test_*.py", "-v"], cwd=root, check=False)
    if result.returncode != 0:
        return result.returncode
    if args.python_only:
        print("Python tests completed. .NET build and integration tests were NOT executed.", flush=True)
        return 0
    dotnet = shutil.which("dotnet")
    if dotnet is None:
        print("The .NET SDK is unavailable. .NET validation did not run.", file=sys.stderr)
        return 2
    project = root / "implementations/dotnet/Tests/Multiplexed.AI.Tests/Multiplexed.AI.Tests.csproj"
    environment = dict(os.environ)
    # Opt in with this exact interpreter. Neither production profiles nor host configuration are changed.
    environment["MULTIPLEXED_PYTHON_EXECUTABLE"] = str(Path(sys.executable).absolute())
    with tempfile.TemporaryDirectory(prefix="hosted-python-validation-") as output:
        result = subprocess.run([dotnet, "test", str(project), "--filter",
                                 "FullyQualifiedName~Multiplexed.AI.Tests.Runtime.Invocation.Workers.Python",
                                 "--logger", "trx;LogFileName=hosted-python.trx", "--results-directory", output],
                                cwd=root, env=environment, check=False)
        if result.returncode != 0:
            return result.returncode
        path = Path(output, "hosted-python.trx")
        if not path.is_file():
            print("No fresh .NET test report was produced; validation failed.", file=sys.stderr)
            return 2
        document = ET.parse(path).getroot()
        counters = document.find(".//{*}Counters")
        if counters is None:
            print("No .NET test counters were returned.", file=sys.stderr)
            return 2
        total = int(counters.get("total", "0"))
        passed = int(counters.get("passed", "0"))
        executed = int(counters.get("executed", "0"))
        required = ("Published_Synchronous_Function_Returns_Real_Computed_Data",
                    "Unstarted_Python_Function_Executes_Original_Code_After_Republication",
                    "Python_Exception_Retains_Uncertain_Invocation_Without_Manufacturing_Business_Failure")
        names = [r.get("testName", "") for r in document.findall(".//{*}UnitTestResult") if r.get("outcome") == "Passed"]
        if total <= 0 or passed != total or executed != total or not all(any(n in name for name in names) for n in required):
            print("Incomplete .NET/Python coverage: zero tests, skipped tests, failures or missing live scenarios.", file=sys.stderr)
            return 2
        print(f"Validated: {passed} .NET cases passed, including real Python and pinned local DAG scenarios; no skipped cases.", flush=True)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
